### PR TITLE
validate person id in clinical table is in person table

### PR DESCRIPTION
**Pull request by ilovemane**

### CODE REVIEW

File: R/classCdmReference.R [hunk 1]
```diff
@@ -277,6 +277,30 @@ checkPlausibleObservationDates <- function(x, call = parent.frame()) {
 
 
 }
+checkPerson <- function(cdm, tableName, id, call = parent.frame) {
+  if (tableName %in% names(cdm)) {
+    x <- cdm[[tableName]]
+
+    if (isTRUE(nrow(x |> utils::head(1) |> dplyr::collect()) > 0)) {
```
Comment (catalamarti) [inline R/classCdmReference.R:284]: else return()

File: R/classCdmReference.R [hunk 2]
```diff
@@ -277,6 +277,30 @@ checkPlausibleObservationDates <- function(x, call = parent.frame()) {
...
+      x <- x |> dplyr::select(.data$person_id) |>
+        dplyr::distinct() |>
+        dplyr::pull()
+    }
+
+    if (!inherits(x, "tbl")) {
```
Comment (catalamarti) [inline R/classCdmReference.R:290]: this if is not needed

File: R/classCdmReference.R [hunk 3]
```diff
@@ -277,6 +277,30 @@ checkPlausibleObservationDates <- function(x, call = parent.frame()) {
 
 
 }
+checkPerson <- function(cdm, tableName, id, call = parent.frame) {
```
Comment (catalamarti) [inline R/classCdmReference.R:280]: parent.frame()

File: R/validate.R [hunk 4]
```diff
@@ -540,6 +543,24 @@ validateCdmArgument <- function(cdm,
  if (isTRUE(checkPlausibleObservationDates)){
    checkPlausibleObservationDates(cdm$observation_period)
 
+ }
+
+ if (isTRUE(checkPerson)){
+
+   id <- cdm$person |> dplyr::pull(.data$person_id)
```
Comment (catalamarti) [inline R/validate.R:550]: would be nice to have this inside the code

File: R/validate.R [hunk 5]
```diff
@@ -540,6 +543,24 @@ validateCdmArgument <- function(cdm,
...
+   tableName <- c("observation_period","visit_occurrence",
```
Comment (catalamarti) [inline R/validate.R:551]: check all the tables, not have a prior list

Comment (catalamarti) [reply R/validate.R:551]: inside you check if person_id or subject_id are columns and then you check

Comment (catalamarti) [reply R/validate.R:551]: try to run in database side:
cdm$my_table |> dplyr::anti_join(cdm$perosn) |> head(1) |> dplyr::collect()

**Review by catalamarti — CHANGES_REQUESTED** [review]

**Comment by ilovemane** [comment]
@catalamarti  This should be ok now

File: R/classCdmReference.R [hunk 6]
```diff
@@ -277,6 +277,49 @@ checkPlausibleObservationDates <- function(x, call = parent.frame()) {
 
 
 }
+checkPerson <- function(cdm, call = parent.frame()) {
+  id <- cdm$person |> dplyr::pull(.data$person_id)
+
+  # check for table with persion_id or subject_id
```
Comment (catalamarti) [inline R/classCdmReference.R:283]: tableName <- names(cdm)
tableName <- tableName [purrr::map_lgl(tableName, \(x) any(c("person_id", "subject_id") %in% colnames(cdm[[x]])))]

File: R/classCdmReference.R [hunk 7]
```diff
@@ -277,6 +277,49 @@ checkPlausibleObservationDates <- function(x, call = parent.frame()) {
 
 
 }
+checkPerson <- function(cdm, call = parent.frame()) {
```
Comment (catalamarti) [inline R/classCdmReference.R:280]: Could we do it without leaving the database?
```
tables <- names(cdm) |>
  rang::set_names() |>
  purrr::map_lgl(\(x) {
    id <- c("person_id", "subject_id")
    id <- id[id %in% colnames(cdm[[x]])]
    if (length(id) == 1) {
      cdm[[x]] |>
        dplyr::select("person_id" = dplyr::all_of(id)) |>
        dplyr::anti_join(cdm$person, by = "person_id") |>
        utils::head(1) |>
        dplyr::tally() |>
        dplyr::pull() > 0
    } else {
      FALSE
    }
  })
Ann then report the trues? Do you think something like that would work, probably you can fins a way to make the code a bit more elegant.
```

**Review by catalamarti — CHANGES_REQUESTED** [review]

File: R/classCdmReference.R (hunk 7 above)
Comment (ilovemane) [reply R/classCdmReference.R:280]: i checked it to using purrr::map now

**Review by catalamarti — APPROVED** [review]

**Comment by catalamarti** [comment]
thank you very much @ilovemane this is good to go

**Pull request merged**